neqo-transport/src/cc/cubic.rs

```diff
@@ -0,0 +1,182 @@
+use std::time::{Duration, Instant};
+
+use crate::cc::{classic_cc::WindowAdjustment, CWND_MIN, MAX_DATAGRAM_SIZE_F64};
+use neqo_common::qtrace;
+use std::convert::TryFrom;
+
+// CUBIC congestion control
+
+// C is a constant fixed to determine the aggressiveness of window
+// increase  in high BDP networks.
+pub const CUBIC_C: f64 = 0.4;
+// beta_cubic is the CUBIC multiplication decrease factor
+const CUBIC_BETA: f64 = 0.7;
+pub const CUBIC_ALPHA: f64 = 3.0 * (1.0 - 0.7) / (1.0 + 0.7);
+
+pub const CUBIC_BETA_USIZE_QUOTIENT: usize = 7;
+pub const CUBIC_BETA_USIZE_DIVISOR: usize = 10;
+
+/// The fast convergence ratio further reduces the congestion window when a congestion event
+/// occurs before reaching the previous `W_max`.
+pub const CUBIC_FAST_CONVERGENCE: f64 = (1.0 + CUBIC_BETA) / 2.0;
+
+#[derive(Debug)]
+pub struct Cubic {
+    last_max_cwnd: f64,
+    estimated_tcp_cwnd: f64,
+    k: f64,
+    w_max: f64,
+    ca_epoch_start: Option<Instant>,
+    last_phase_was_tcp: bool,
+    tcp_acked_bytes: f64,
+}
+
+impl Default for Cubic {
+    fn default() -> Self {
+        Self {
+            last_max_cwnd: 0.0,
+            estimated_tcp_cwnd: 0.0,
+            k: 0.0,
+            w_max: 0.0,
+            ca_epoch_start: None,
+            last_phase_was_tcp: false,
+            tcp_acked_bytes: 0.0,
+        }
+    }
+}
+
+impl Display for Cubic {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(
+            f,
+            "Cubic [last_max_cwnd: {}, k: {}, w_max: {}, ca_epoch_start: {:?}]",
+            self.last_max_cwnd, self.k, self.w_max, self.ca_epoch_start
+        )?;
+        Ok(())
+    }
+}
+
+#[allow(clippy::doc_markdown)]
+impl Cubic {
+    /// Original equations is:
+    /// K = cubic_root(W_max*(1-beta_cubic)/C) (Eq. 2 RFC8312)
+    /// W_max is number of segments of the maximum segment size (MSS).
+    ///
+    /// K is actually the time that W_cubic(t) = C*(t-K)^3 + W_max (Eq. 1) would
+    /// take to increase to W_max. We use bytes not MSS units, therefore this
+    /// equation will be: W_cubic(t) = C*MSS*(t-K)^3 + W_max.
+    ///
+    /// From that equation we can calculate K as:
+    /// K = cubic_root((W_max - W_cubic) / C / MSS);
+    fn calc_k(&self, curr_cwnd: f64) -> f64 {
+        ((self.w_max - curr_cwnd) / CUBIC_C / MAX_DATAGRAM_SIZE_F64).cbrt()
+    }
+
+    /// W_cubic(t) = C*(t-K)^3 + W_max (Eq. 1)
+    fn w_cubic(&self, t: f64) -> f64 {
+        CUBIC_C * (t - self.k).powi(3) * MAX_DATAGRAM_SIZE_F64 + self.w_max
+    }
+}
+
+impl WindowAdjustment for Cubic {
+    // This is because of the cast in the last line from f64 to usize.
+    #[allow(clippy::cast_possible_truncation)]
+    #[allow(clippy::cast_sign_loss)]
+    fn on_packets_acked(
+        &mut self,
+        curr_cwnd: usize,
+        acked_bytes: usize,
+        min_rtt: Duration,
+        now: Instant,
+    ) -> usize {
+        let curr_cwnd_f64 = f64::try_from(u32::try_from(curr_cwnd).unwrap()).unwrap();
```

Comment:
I was thinking about add such a conversion function, but I do not really think we can overflow u32. I expect that we will hit some other limits, e.g. CPU, socket buffers, etc. (we will hit this limit with something like 8GB * 250ms)

... ok, I will add the conversion. The other workaround this problem that I could think of was to keep curr_cwnd separate for u64 and f64 to avoid the conversion. The curr_cwnd is increase and decrease in one place so it would not be hard to keep them sync.... this is not worth the effort.